### PR TITLE
Braz-5: Create user endpoint

### PR DESCRIPTION
**Pull request by mhmBraz**

### CODE REVIEW

**Review by italoservio — CHANGES_REQUESTED** [review]

File: services/users/infra/http/user_validator_controller.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	myValidator := &XValidator{
+		validator: validate,
+	}
+
+	if errs := myValidator.Validate(payloadValidate, validate); len(errs) > 0 && errs[0].Error {
+		return errors.New(exception.CodeInternal)
+	}
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply services/users/infra/http/user_validator_controller.go]: https://github.com/italoservio/braz_ecommerce/pull/13/commits/7780cc3833a28127e4375285a0bff8680c256ffa

File: services/users/infra/http/user_controller_test.go [hunk 2]
```diff
@@ -173,3 +179,153 @@ func TestUserController_DeleteUserById(t *testing.T) {
...
+
+		var httpResponse exception.HTTPException
+		json.Unmarshal(bytes, &httpResponse)
+
+		assert.Equal(t, 400, httpResponse.StatusCode, "should return expected status code")
+		assert.Equal(t, "Invalid input for one or more required attributes", httpResponse.ErrorMessage, "should return expected error message")
+	})
+
+	t.Run("should mount the http exception when there is an error in ValidationRequest", func(t *testing.T) {
+		bodyReader := strings.NewReader(`{"Username": "12124", "Password": "testinasg", "Channel": "M"}`)
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply services/users/infra/http/user_controller_test.go]: https://github.com/italoservio/braz_ecommerce/pull/13/commits/3900a0a786cb79635c675464838bf9dfaa27ad70

File: services/users/infra/http/user_controller_test.go [hunk 3]
```diff
@@ -173,3 +179,153 @@ func TestUserController_DeleteUserById(t *testing.T) {
...
+
+		var httpResponse exception.HTTPException
+		json.Unmarshal(bytes, &httpResponse)
+
+		assert.Equal(t, 500, httpResponse.StatusCode, "should return expected status code")
+		assert.Equal(t, "Failed to communicate with database", httpResponse.ErrorMessage, "should return expected error message")
+	})
+
+	t.Run("should return empty error when successfully executed on ValidationRequest and createUser", func(t *testing.T) {
+		bodyReader := strings.NewReader(`{"first_name": "testinasg", "last_name": "testinasg", "email": "testinasg", "type": "testinasg", "password": "testinasg"}`)
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply services/users/infra/http/user_controller_test.go]: https://github.com/italoservio/braz_ecommerce/pull/13/commits/3900a0a786cb79635c675464838bf9dfaa27ad70

File: services/users/infra/http/user_controller_test.go [hunk 4]
```diff
@@ -173,3 +179,153 @@ func TestUserController_DeleteUserById(t *testing.T) {
...
+
+		var httpResponse exception.HTTPException
+		json.Unmarshal(bytes, &httpResponse)
+
+		assert.Equal(t, 400, httpResponse.StatusCode, "should return expected status code")
+		assert.Equal(t, "Invalid input for one or more required attributes", httpResponse.ErrorMessage, "should return expected error message")
+	})
+
+	t.Run("should mount http exception when receiving an error from app", func(t *testing.T) {
+		bodyReader := strings.NewReader(`{"first_name": "testinasg", "last_name": "testinasg", "email": "testinasg", "type": "testinasg", "password": "testinasg"}`)
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply services/users/infra/http/user_controller_test.go]: https://github.com/italoservio/braz_ecommerce/pull/13/commits/3900a0a786cb79635c675464838bf9dfaa27ad70

File: services/users/app/create_user.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	database.DatabaseTimestamp `bson:",inline"`
+}
+
+func (gu *CreateUserImpl) Do(createUser *CreateUserInput) (*CreateUserOutput, error) {
+	secret := os.Getenv("ENC_SECRET")
+	encryptionData, err := encryption.Encrypt(secret, createUser.Password)
+
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply services/users/app/create_user.go]: https://github.com/italoservio/braz_ecommerce/pull/13/commits/46bf56b3e2e6795b83d8cc2cec8d0f830009378b

File: packages/exception/http_exception.go [hunk 6]
```diff
@@ -57,6 +61,14 @@ func errorCodeToStruct(code string) *HTTPException {
...
 		response.StatusCode = http.StatusBadRequest
 		response.ErrorMessage = "Invalid input for one or more required attributes"
+	case CodeInternal:
+		response.StatusMessage = "Internal Server Error"
+		response.StatusCode = http.StatusInternalServerError
+		response.ErrorMessage = "An expected error occurred and the server could not deal with it"
+	case CodeValidationBodyFailed:
+		response.StatusMessage = "Body Error"
+		response.StatusCode = http.StatusBadRequest
+		response.ErrorMessage = "Invalid input for one or more required attributes"
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply packages/exception/http_exception.go]: https://github.com/italoservio/braz_ecommerce/pull/13/commits/ac0ee704ea8cb45169a0610e710ee65cc1c4ca8f

File: packages/encryption/encrypt.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+type EncryptedText struct {
+	EncryptedText string
+	Salt          string
+}
+
+func Encrypt(secret string, text string) (*EncryptedText, error) {
+	if secret == "" || text == "" {
+		return nil, errors.New(exception.CodeValidationFailed)
+	}
```
[the remark this answers is not in the document]
Comment (mhmBraz) [reply packages/encryption/encrypt.go]: https://github.com/italoservio/braz_ecommerce/pull/13/commits/d8e9e64c1308baf013bc8d2948bd5d234ec0ec19

**Review by italoservio — APPROVED** [review]

**Pull request merged**